### ISSUE TITLE
Consider order of mixed tag and branch elements in configuration

### ISSUE THREAD

**Issue by yankydoo** [qoomon/maven-git-versioning-extension#125]
Suppose we have the following configuration:

```
<configuration xmlns="https://github.com/qoomon/maven-git-versioning-extension" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
               xsi:schemaLocation="https://github.com/qoomon/maven-git-versioning-extension https://qoomon.github.io/maven-git-versioning-extension/configuration-6.4.0.xsd">
    <branch>
         <pattern><![CDATA[maintenance/(?<branchVersion>[0-9\-\.]+)]]></pattern>
         <versionFormat>${branchVersion}.x-MAINTENANCE-SNAPSHOT</versionFormat>
    </branch>
    <tag>
	      <pattern><![CDATA[(?<tagVersion>[0-9].*)]]></pattern>
        <versionFormat>${tagVersion}</versionFormat>
    </tag>
    <branch>
         <pattern><![CDATA[master/(?<branchVersion>[0-9\-\.]+)]]></pattern>
         <versionFormat>${branchVersion}.x-SNAPSHOT</versionFormat>
    </branch>
</configuration>
```

Following situations and our desired outcome:

1.  branch is maintenance/1.0 and tag is 1.0.0
  => version 1.0.x-MAINTENANCE-SNAPSHOT (because the maintenance pattern matches and it is before the tag configuration)
2.  branch is master/1.0 and tag is 1.0.0
  => version 1.0.0 (because the tag pattern matches and it is before the master branch configuration)
3.  branch is master/1.0 and no tag
  => version is 1.0.x-SNAPSHOT

Basically, we'd like that the ordering of the tag and branch elements in the configuration is considered.

Advantages from our POV:
- preferTags would not be needed anymore (if you prefer it, just put it at the top)
- the configuration would allow greater flexibility

Would you be open to such a change? Do you see any downside to this?

Offtopic to this issue, I also think some strict mode would be great, e.g., do not fall back to the DEFAULT formats in the code but rather abort the build.

**Comment by qoomon** [comment]
HI @yankydoo, I like both ideas.

Regarding the strict mode, I think I'll just remove defaults in general.

I'll keep you posted as soon as I have a running version. Probably within the next days.

**Comment by qoomon** [comment]
currently I'm thinking about changing `<branch> <tag> <commit>` configs to something like 
```xml
<ref>
  <type>branch</type>
  <pattern>main</pattern>
  <versionFormat>${version}</versionFormat>
</ref>
```
or maybe 
```xml
<ref type="branch">
  <pattern>main</pattern>
  <versionFormat>${version}</versionFormat>
</ref>
```
WDYT?

**Comment by qoomon** [comment]
Currently I have some concern regarding performance, if I always check for tags pointing at current commit.
Maybe we still need an option to do so.

Another issue I have is what is the expected behavior in detached head state. Should I determine all branches for this commit and try to find the first matching one?

**Comment by qoomon** [comment]
current behavior
* if on branch 
  * -> use matching branch config
 
* if in detached head state
  * get all tags (~1sec) 
  * if tags point on commit
    * -> use matching tag config
    
* -> use commit config

could you describe your expected behavior alike?

**Comment by yankydoo** [comment]
> currently I'm thinking about changing <branch> <tag> <commit> configs to something like
> ...
> WDYT?

I have no strong opinions on the syntax. IMO you could also keep it as-is but I guess it's easier for you to parse `<ref type=...>` and it also would still be correct. Changing the syntax would be (another) breaking change - it's up to you :)

> could you describe your expected behavior alike?

Our desired behavior would be

* if on branch
  * get all tags
  * -> use first matching branch or tag config
* if in detached head state
  * get all tags (~1sec)
  * if tags point on commit
    * -> use matching tag config 
* -> use commit config

> Another issue I have is what is the expected behavior in detached head state. Should I determine all branches for this commit and try to find the first matching one?

In detached head state I just wouldn't consider the branch configs at all (I can't really think of a use case for that behavior)

> Currently I have some concern regarding performance, if I always check for tags pointing at current commit.
> Maybe we still need an option to do so.

Generally, I think that performance when getting the tags is not a big concern - when preferTags is active this is already the case and at least for us it's no problem. If you think it is a problem we wouldn't mind an additional option either (something like considerTagsOnBranches) - the preferTags option could still be removed in this case (superceded by ordering of refs).

**Comment by qoomon** [comment]
sound reasonable.

I think I'll go for following notation
```xml
<refs>
  <ref branch="main">
    <versionFormat>${version}</versionFormat>
  </ref>
</refs>
```

**Comment by qoomon** [comment]
maybe it's too complicated :-/, I'm not sure

**Comment by qoomon** [comment]
:-D Currently I'm thinking of leaving configuration as is and just add the option `considerTagsOnBranches`.

It would be downward compatible with `prefereTags` option, although I would mark that option as deprecated.

**Comment by yankydoo** [comment]
> 
> 
> :-D Currently I'm thinking of leaving configuration as is and just add the option `considerTagsOnBranches`.
> 
> It would be downward compatible with `prefereTags` option, although I would mark that option as deprecated.

That sounds like a good solution.

Thank you for being so responsive to suggestions. It's a pleasure to work with you :)

**Comment by qoomon** [comment]
```xml
<configuration xmlns="https://github.com/qoomon/maven-git-versioning-extension" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
               xsi:schemaLocation="https://github.com/qoomon/maven-git-versioning-extension https://qoomon.github.io/maven-git-versioning-extension/configuration-7.0.0.xsd">

    <considerTagsAtBranch>true</considerTagsAtBranch>

    <refs>
        <ref type="tag">
            <pattern>version/(.*)</pattern>
            <version>${1}</version>
        </ref>
        <ref type="branch">
            <pattern>master(-.+)?</pattern>
            <version>${version.release}-master</version>
            <properties>
                <foo>new_new_new</foo>
            </properties>
        </ref>
    </refs>

    <rev>
        <version>${1}</version>
    </rev>

</configuration>
```

**Comment by yankydoo** [comment]
I'd be okay with that but I thought you didn't want to change the existing schema (branch/tag/commit vs refs/rev)? Or did I misunderstand?

**Comment by qoomon** [comment]
You are totally right, I've changed my mind :-). Multiple reasons for that
* I don't like the fact that you can mix `rev` and `refs` configs
* It's hard to parse `branch` and `tag` element in a way I need it 
* I simplified config: `versionFormat` to just `version` same for `properties` this now matches the pom field/elements names
* This will break current configs so everybody needs to understand the new one

**Comment by qoomon** [comment]
Implementation is done, however I still need to patch some tests.

The only issue I have is that I'm not sure about the configuration wordings.
* I don't think everybody is familiar with `ref` and `rev`
  * What about renaming `rev` to `default`?
* Do you have any idea how to rename `refs` and `ref`

**Comment by qoomon** [comment]
just an idea
```xml
<configuration>

    <considerTagsAtBranch>true</considerTagsAtBranch>

    <switch>
        <tag>
            <pattern>version/(.*)</pattern>
            <version>${1}</version>
        </tag>
        <branch>
            <pattern>master(-.+)?</pattern>
            <version>${version.release}-master</version>
            <properties>
                <foo>new_new_new</foo>
            </properties>
        </branch>
        <default>
            <version>${1}</version>
        </default>
    </switch>

</configuration>
```

**Comment by qoomon** [comment]
or maybe 
```xml
<configuration>

    <considerTagsAtBranch>true</considerTagsAtBranch>

    <switch>
        <case ref="tag">
            <pattern>version/(.*)</pattern>
            <version>${1}</version>
        </case>
        <case ref="branch">
            <pattern>master(-.+)?</pattern>
            <version>${version.release}-master</version>
            <properties>
                <foo>new_new_new</foo>
            </properties>
        </case>
        <default>
            <version>${1}</version>
        </default>
    </switch>

</configuration>
```

**Comment by qoomon** [comment]
current iteration
```xml
<configuration xmlns="https://github.com/qoomon/maven-git-versioning-extension" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
               xsi:schemaLocation="https://github.com/qoomon/maven-git-versioning-extension https://qoomon.github.io/maven-git-versioning-extension/configuration-7.0.0.xsd">

    <describeTagPattern>v.*</describeTagPattern>
    
    <refs considerTagsAtBranch="true">
        
        <ref type="tag">
            <pattern>version/(.*)</pattern>
            <version>${1}</version>
            <updatePom>true</updatePom>
        </ref>

        <ref type="branch">
            <pattern>master</pattern>
            <version>${version.release}-${branch}</version>
            <properties>
                <foo>new_new_new</foo>
            </properties>
        </ref>
        
        <ref type="branch">
            <pattern>feature/(.+)</pattern>
            <version>${version.release}-master</version>
            <properties>
                <foo>new_new_new</foo>
            </properties>
        </ref>
        
        <ref type="commit">
            <version>${commit.short}</version>
        </ref>
        
    </refs>
    
</configuration>

```

**Comment by qoomon** [comment]
I am thinking about to change the default behaviour to always consider tags on branches and add an option to disable this behaviour on demand e.g. `<refs considerTagsOnlyIfHeadIsDetached="true">` WDYT?

**Comment by qoomon** [comment]
@yankydoo If you like to try the new version just checkout https://github.com/qoomon/maven-git-versioning-extension/tree/release/7.0.0

**Comment by yankydoo** [comment]
Sorry for the late response. I like your last iteration and also think that it's a good idea that the default behavior is to consider tags.

Regarding `<ref type="commit">`:  Maybe you could add an XSD restriction to enforce the implicit rules (at most one/should be last one), or alternatively put out a warning.

I'll give it a go this week!

**Comment by qoomon** [comment]
@yankydoo with current configuration xml structure it is impossible to implement a validation we'd like to have. (at least with xsd 1.0)

I tend to following configuration again :-) and to mark rev config as required to always be able build a project.
Or I change the behaviour in such a way that if no ref config is provided the pom xml is used as is, however I think a mandatory rev config is a better way.
```xml
<configuration xmlns="https://github.com/qoomon/maven-git-versioning-extension"
               xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
               xsi:schemaLocation="https://github.com/qoomon/maven-git-versioning-extension https://qoomon.github.io/maven-git-versioning-extension/configuration-7.0.0.xsd">

    <describeTagPattern><![CDATA[v(?<version>.*)]]></describeTagPattern>

    <rev>
        <version>${describe.tag.version}-${describe.distance}-${commit.short}</version>
    </rev>
    
    <refs considerTagsAtBranch="true">

        <ref type="tag">
            <pattern><![CDATA[v(?<version>.*)]]></pattern>
            <version>${ref.version}</version>
            <updatePom>true</updatePom>
        </ref>

        <ref type="branch">
            <pattern>master</pattern>
            <version>${project.version.release}-${branch}</version>
            <properties>
                <foo>new_new_new</foo>
            </properties>
        </ref>

        <ref type="branch">
            <pattern>master</pattern>
            <version>${project.version.release}-${branch}</version>
            <properties>
                <foo>new_new_new</foo>
            </properties>
        </ref>

        <ref type="branch">
            <pattern>feature/(.+)</pattern>
            <version>${ref}-SNAPSHOT</version>
            <properties>
                <foo>new_new_new</foo>
            </properties>
        </ref>
    </refs>

</configuration>
```

**Comment by qoomon** [comment]
@yankydoo  I'm done with developing this release and all tests are running successfully again. I'll wait for some feedback from your side and then I'll eventually do the release of `7.0.0`

**Comment by qoomon** [comment]
@yankydoo I think I have a final version, could you give some feedback regarding functionality, readme and changelog. looking forward.

**Comment by yankydoo** [comment]
> 
> 
> @yankydoo I think I have a final version, could you give some feedback regarding functionality, readme and changelog. looking forward.

Sorry for the late response. I'm a little busy at the moment but I'll have a look at it this weekend.

**Comment by yankydoo** [comment]
I just gave it a go and it's looking good - our use case works 🎉

README and CHANGELOG also look good (all the breaking changes I stumbled across are documented). 

I noticed the new CI/CD defaults and at least in our Jenkins setup we use CHANGE_BRANCH (if available) over BRANCH_NAME. The reason is that in Pull Requests we want the build to use the name of the source branch (so CHANGE_BRANCH). As long as setting `VERSIONING_GIT_BRANCH` overrides the default it should work for us.

**Comment by qoomon** [comment]
glad to hear everything is working as expected. regarding CI/CD defaults yes if you set one of `VERSIONING_GIT_...` environment variables CI/CD defaults won't be considered. Does it make sense to use `CHANGE_BRANCH` value if present? I'm not used to jenkins, so some smart suggestions are appreciated :-)

**Comment by qoomon** [comment]
maybe something like `$CHANGE_BRANCH/$BRANCH_NAME`, if `$CHANGE_BRANCH` is set. This would lead to branches like `PR-123/feature/skyrocket`

**Comment by qoomon** [comment]
I did some research regarding pull requests I'll leave the CI/CD feature as it is right now. Everybody will be able to customize branch name by setting `VERSIONING_GIT_...` variables.

I'll probably will release the new version (maven and gradle) today.

**Comment by qoomon** [comment]
Version `5.0.0` has been released.
